vowpalwabbit/interactions.h

```diff
@@ -28,49 +34,113 @@ inline constexpr bool is_printable_namespace(const unsigned char ns)
 
-// remove duplicate interactions and sort namespaces in them (if required)
 void sort_and_filter_duplicate_interactions(
     std::vector<std::vector<namespace_index>>& vec, bool filter_duplicates, size_t& removed_cnt, size_t& sorted_cnt);
 
-/*
- *  Feature combinations generation
- */
+template <generate_func_t generate_func, bool leave_duplicate_interactions>
+std::vector<std::vector<namespace_index>> compile_interaction(
+    const std::vector<namespace_index>& interaction, const std::set<namespace_index>& indices)
+{
+  std::vector<size_t> insertion_indices;
+  std::vector<namespace_index> insertion_ns;
+  size_t num_wildcards = 0;
+  for (size_t i = 0; i < interaction.size(); i++)
+  {
+    if (interaction[i] != wildcard_namespace)
+    {
+      insertion_indices.push_back(i);
+      insertion_ns.push_back(interaction[i]);
+    }
+    else
+    {
+      num_wildcards++;
+    }
+  }
 
-// function estimates how many new features will be generated for example and their sum(value^2).
-void eval_count_of_generated_ft(bool permutations, const std::vector<std::vector<namespace_index>>& interactions,
-    const std::array<features, NUM_NAMESPACES>& feature_spaces, size_t& new_features_cnt, float& new_features_value);
+  // Quadratic fast path or generic generation function.
+  auto result = num_wildcards == 2 ? expand_quadratics_wildcard_interactions(leave_duplicate_interactions, indices)
+                                   : generate_func(indices, num_wildcards);
+  for (size_t i = 0; i < insertion_indices.size(); i++)
+  {
+    for (auto& res : result) { res.insert(res.begin() + insertion_indices[i], insertion_ns[i]); }
+  }
+  return result;
+}
 
-template <class R, class S, void (*T)(R&, float, S), bool audit, void (*audit_func)(R&, const audit_strings*)>
-inline void generate_interactions(vw& all, example_predict& ec, R& dat, size_t& num_interacted_features)
+template <generate_func_t generate_func, bool leave_duplicate_interactions>
+std::vector<std::vector<namespace_index>> compile_interactions(
+    const std::vector<std::vector<namespace_index>>& interactions, const std::set<namespace_index>& indices)
 {
-  if (all.weights.sparse)
-    generate_interactions<R, S, T, audit, audit_func, sparse_parameters>(
-        *ec.interactions, all.permutations, ec, dat, all.weights.sparse_weights, num_interacted_features);
-  else
-    generate_interactions<R, S, T, audit, audit_func, dense_parameters>(
-        *ec.interactions, all.permutations, ec, dat, all.weights.dense_weights, num_interacted_features);
+  std::vector<std::vector<namespace_index>> final_interactions;
+
+  for (const auto& inter : interactions)
+  {
+    if (contains_wildcard(inter))
+    {
+      auto compiled = compile_interaction<generate_func, leave_duplicate_interactions>(inter, indices);
+      std::copy(compiled.begin(), compiled.end(), std::back_inserter(final_interactions));
```

Comment:
any way to move these vectors instead of copying them? Does this fn get called enough for it to matter?

Reply:
In these functions the intention is to rely on RVO and NRVO rather than explicitly moving. Often times explicit move calls actually prevents RVO and results in worse performance. The number of times these are called should be infrequent enough to not matter though.